tests/algorithm_validation_tests/test_one_way_anova.py

```diff
@@ -71,7 +71,44 @@ def test_anova_algorithm(test_input, expected):
     for key, e_val in e_aov.items():
         r_val = aov[key]
         assert e_val == r_val or np.isclose(e_val, r_val)
-    for et, rt in zip(e_tukey, tukey):
-        for key, e_val in et.items():
-            r_val = rt[key]
-            assert e_val == r_val or np.isclose(e_val, r_val, rtol=1e-7, atol=1e-10)
+    compare_results(e_tukey, tukey)
+
+
+def compare_results(expected, result):
+    assert len(expected) == len(result)
+    for grp_res_a in expected:
+        groupA_exp = grp_res_a["groupA"]
+        groupB_exp = grp_res_a["groupB"]
+        for grp_res_b in result:
+            groupA_res = grp_res_b["groupA"]
+            groupB_res = grp_res_b["groupB"]
+            if groupA_exp == groupA_res and groupB_exp == groupB_res:
+                compare_as_is(grp_res_a, grp_res_b)
+                break
+            elif groupA_exp == groupB_res and groupB_exp == groupA_res:
+                compare_opposites(grp_res_a, grp_res_b)
+                break
+        else:
+            raise Exception(
+                f"Group categories {groupA_exp} - {groupB_exp} could not be found in {result}"
+            )
+
+
+def compare_as_is(expected, result):
+    for key, e_val in expected.items():
+        r_val = result[key]
+        if isinstance(e_val, str) and isinstance(r_val, str):
+            assert e_val == r_val
+        else:
+            assert np.isclose(e_val, r_val, rtol=1e-7, atol=1e-10)
+
+
+def compare_opposites(expected, result):
+    assert expected["groupA"] == result["groupB"]
+    assert expected["groupB"] == result["groupA"]
+    assert np.isclose(expected["meanA"], result["meanB"], rtol=1e-7, atol=1e-10)
+    assert np.isclose(expected["meanB"], result["meanA"], rtol=1e-7, atol=1e-10)
+    assert np.isclose(expected["diff"], result["diff"] * -1, rtol=1e-7, atol=1e-10)
+    assert np.isclose(expected["se"], result["se"], rtol=1e-7, atol=1e-10)
+    assert np.isclose(expected["t_stat"], result["t_stat"] * -1, rtol=1e-7, atol=1e-10)
+    assert np.isclose(expected["p_tuckey"], result["p_tuckey"], rtol=1e-7, atol=1e-10)
```

Comment:
This is very hard to understand. 

Assuming `res` means result and `exp` means expected. You are creating `grp_res_a` from `expected`, `groupA_exp` from `grp_res_a`, `groupB_exp` from `grp_res_a` and `groupA_res` from `grp_res_b`. These are all inconsistent with respect to both the `res`/`exp` component and the `a`/`b` one. 

Then, why do you need the double loop? If you could make sure that your implementation returns the groups **in the same order** as the SOA implementation, you would just zip `result` and `expected` and make a single pass. I am not saying this for performance, it would just make things easier to understand. 

Continuing. We said in our call that iterating over all items and using a generic assertion is bad because when something fails it is impossible to see what was it. You create two functions, `compare_as_is` and `compare_opposites` to replace the original piece of code. The second on **correctly has one assertions per quantity** but the first one just repeats the original logic. Why? 

Then, why do you even need two functions? I understand that, since you didn't make sure that you get the results in the same order as in the SOA implementation, some quantities have inverted signs. That's OK since what really matters is not the order of the groups nor the sign, but the magnitude of the given quantity. You can perfectly well construct appropriate assertions for these conditions:
```python
assert {groupA_res, groupB_res} == {groupA_exp, groupB_exp}  # ignore order
assert isclose(x, y)  # when the sign matters
assert isclose(abs(x), abs(y))  # when the sign doesn't matter 
```
additionally, since `diff` and `t_stat` always come with the same sign, you can add one more assertion for consistency
```python
assert sign(diff) == sign(t_stat)
```
The above assertions make clear exactly what property is desired.

And two last things. 

I told you in past reviews not to catch a base `Exception` but to always use a [concrete exception](https://docs.python.org/3/library/exceptions.html#concrete-exceptions). The same is true for raising. If you don't know which one, use `ValueError`. 

 There is a function in `tests.algorithm_validation_tests.helpers` called `assert_all_close`. We should all use this one to have the same tolerance in all tests. Besides, it uses `numpy.testing.assert_allclose` which is designed for testing and gives better feedback than `numpy.isclose`.

Reply:
I thought of creating something that reflects exactly the mip-engine implementation structure, however, if anything changes for any reason in the implementation, then we should also make changes in the tests and I think we shouldn't have to come to that. 

I though of a simple way to do what needed to be done, since there were a lot of issues at the time.

Our call occurred several days after creating this PR, in the paired t-test I actually used the helper methods. 

> The second on **correctly has one assertions per quantity** but the first one just repeats the original logic. Why?

Because, our conversation occurred after creating this PR I prioritised closing other more important matters. I will be using the new helper functions and have these suggestions in mind, while creating a ticket with low priority to refactor these test. For now, if we don't consider this a major issue, I will merge this.